### PR TITLE
Issue 2397: Enable unqualified stream names as part of ReaderGroupConfig.

### LINKED ISSUE #2397: ReaderGroupConfig should allow unqualified stream names.
**Problem description**
The current `ReaderGroupConfig` supports the the following two ways of passing a stream.
```
1. ReaderGroupConfig.builder().stream("scope/s1", startStreamCut1).build() // full scoped streamName
2. ReaderGroupConfig.builder().stream(stream1) // stream object.
              .stream(stream2, startStreamCut2).build() 
```
 It should also allow unqualified stream names e.g: (builder.stream("stream2")), i.e. use the RG scope by default.

**Problem location**
ReaderGroupConfig.

**Suggestions for an improvement**
Use RG Scope by default.
ref: https://github.com/pravega/pravega/pull/2374#pullrequestreview-105047939

### PR DESCRIPTION
**Pull request by shrids**
**Change log description**
Enable use of unqualified stream names as part of ReaderGroupConfig. Use the Reader Group's Scope by default in such a scenario.

**Purpose of the change**
Fixes #2397

**What the code does**
* Enable unqualified stream names. 
```
//ReaderGroup's scope is used by default in the below ReaderGroup configurations.
- ReaderGroupConfig.builder().stream("streamA", startStreamCut1).build(); 
- ReaderGroupConfig.builder().stream("streamB", startStreamCutB, endStreamCutB).build();
```
* Ensure the stream name and scope are passed as separate parameters
```
ReaderGroupConfig.builder()
                 .stream("scopeA", "streamA")
                 .stream("scopeB", "streamB", startStreamCutB)
                 .stream("scopeC", "streamC", startStreamCutC, endStreamCutC)
                 .build();
```

**How to verify it**
All existing tests should pass, new tests have been added to cover the new API

### CODE REVIEW

File: client/src/main/java/io/pravega/client/stream/ReaderGroupConfig.java [hunk 1]
```diff
@@ -58,30 +59,30 @@ public ReaderGroupConfigBuilder disableAutomaticCheckpoints() {
...
         * @return Reader group config builder.
         */
        public ReaderGroupConfigBuilder stream(final String scopedStreamName, final StreamCut startStreamCut, final StreamCut endStreamCut) {
-           final Stream stream = Stream.of(scopedStreamName);
 
            if (startingStreamCuts == null) {
                startingStreamCuts = new HashMap<>();
            }
-           this.startingStreamCuts.put(stream, startStreamCut);
+           this.startingStreamCuts.put(scopedStreamName, startStreamCut);
```
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/stream/ReaderGroupConfig.java]: This change results in a confusing state of affairs: "scopedStreamName" may or may not be scoped, and as such the stream names in `startingStreamCuts` are also maybe or maybe not scoped. We should find a cleaner way to do this.

File: client/src/main/java/io/pravega/client/segment/impl/SegmentInputStreamImpl.java [hunk 2]
```diff
@@ -28,7 +28,7 @@
 import static io.pravega.client.segment.impl.EndOfSegmentException.ErrorType.END_OFFSET_REACHED;
 
 /**
- * Manages buffering and provides a synchronous to {@link AsyncSegmentInputStream}
+ * Manages buffering and provides a synchronus to {@link AsyncSegmentInputStream}
```
Comment (fpj) [inline client/src/main/java/io/pravega/client/segment/impl/SegmentInputStreamImpl.java]: Revert this change, please.

File: client/src/main/java/io/pravega/client/stream/ReaderGroupConfig.java (hunk 1 above)
Comment (fpj) [reply client/src/main/java/io/pravega/client/stream/ReaderGroupConfig.java]: The problem is that there is some implicit behavior, which is prone to confusion and error. Perhaps one solution would be to have two constructors such that one takes both scope and stream names, while the second only takes the stream name only. The latter uses the RG scope as the default one. This way we do not make the behavior ambiguous. Would it work?

Comment (shrids) [reply client/src/main/java/io/pravega/client/stream/ReaderGroupConfig.java]: I have updated the code such that the user can pass either scope and stream name or just the stream name.

File: client/src/main/java/io/pravega/client/segment/impl/SegmentInputStreamImpl.java (hunk 2 above)
Comment (shrids) [reply client/src/main/java/io/pravega/client/segment/impl/SegmentInputStreamImpl.java]: done.

File: client/src/main/java/io/pravega/client/stream/ReaderGroupConfig.java [hunk 3]
```diff
@@ -58,46 +61,95 @@ public ReaderGroupConfigBuilder disableAutomaticCheckpoints() {
         * Add a stream and its associated start {@link StreamCut} and end {@link StreamCut} to be read by the
         * readers of a ReaderGroup.
         *
-        * @param scopedStreamName Scoped Name of the Stream.
-        * @param startStreamCut Start {@link StreamCut}
-        * @param endStreamCut End {@link StreamCut}
+        * @param scope Scope of the stream, if it is not specified then {@link ReaderGroup#getScope()} is used for the scope.
```
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/stream/ReaderGroupConfig.java]: Does "Not specified" mean null? if so just say "null"

Given that this method is overloaded, I don't think we need to document what if it is not specified, they can simply use the other method. This method might even disallow passing null.

File: client/src/main/java/io/pravega/client/stream/ReaderGroupConfig.java [hunk 4]
```diff
@@ -58,46 +61,95 @@ public ReaderGroupConfigBuilder disableAutomaticCheckpoints() {
...
         */
-       public ReaderGroupConfigBuilder stream(final String scopedStreamName, final StreamCut startStreamCut, final StreamCut endStreamCut) {
-           final Stream stream = Stream.of(scopedStreamName);
+       public ReaderGroupConfigBuilder stream(final String scope, final String streamName,
+                                              final StreamCut startStreamCut, final StreamCut endStreamCut) {
+
+           Exceptions.checkNotNullOrEmpty(streamName, "streamName");
+
+           final String scopedStreamName;
+           if (StringUtils.isBlank(scope)) {
```
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/stream/ReaderGroupConfig.java]: Why isBlank()? This seems like it's asking for trouble. What if someone has a scope that is just whitespace or ""?

File: client/src/main/java/io/pravega/client/stream/ReaderGroupConfig.java [hunk 5]
```diff
@@ -34,8 +37,8 @@
     @Getter
     private final long automaticCheckpointIntervalMillis;
 
-    private final Map<Stream, StreamCut> startingStreamCuts;
-    private final Map<Stream, StreamCut> endingStreamCuts;
+    private final Map<String, StreamCut> startingStreamCuts;
```
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/stream/ReaderGroupConfig.java]: This still has the same problem as before. 
If they aren't passing the scope and you are inferring it, it should do so and then instantiate the `Stream` object. We shouldn't have an object whos type/content is ambiguous.

File: client/src/main/java/io/pravega/client/stream/ReaderGroupConfig.java (hunk 4 above)
Comment (fpj) [reply client/src/main/java/io/pravega/client/stream/ReaderGroupConfig.java]: That's a good question, but is this something we want to support? Either way, we need to be consistent all around. Please make sure that the naming is consistent across all of our API. In fact, that's one reason we have `NameUtils` under `shared`.

File: client/src/main/java/io/pravega/client/stream/ReaderGroupConfig.java [hunk 6]
```diff
@@ -58,46 +61,95 @@ public ReaderGroupConfigBuilder disableAutomaticCheckpoints() {
...
+        * @param streamName Stream name.
+        * @param startStreamCut Start {@link StreamCut}.
+        * @param endStreamCut End {@link StreamCut}.
         * @return Reader group config builder.
         */
-       public ReaderGroupConfigBuilder stream(final String scopedStreamName, final StreamCut startStreamCut, final StreamCut endStreamCut) {
-           final Stream stream = Stream.of(scopedStreamName);
+       public ReaderGroupConfigBuilder stream(final String scope, final String streamName,
+                                              final StreamCut startStreamCut, final StreamCut endStreamCut) {
+
```
Comment (fpj) [inline client/src/main/java/io/pravega/client/stream/ReaderGroupConfig.java]: I think we shouldn't have such an extra line according to our style guidelines.

File: client/src/main/java/io/pravega/client/stream/ReaderGroupConfig.java [hunk 7]
```diff
@@ -58,46 +61,95 @@ public ReaderGroupConfigBuilder disableAutomaticCheckpoints() {
...
+        * Add a stream and its associated start {@link StreamCut} and end {@link StreamCut} to be read by the
+        * readers of a ReaderGroup. {@link ReaderGroup#getScope()} is used for the scope of the stream.
+        *
+        * @param streamName Stream name.
+        * @param startStreamCut Start {@link StreamCut}.
+        * @param endStreamCut End {@link StreamCut}.
+        * @return Reader group config builder.
+        */
+       public ReaderGroupConfigBuilder stream(final String streamName, final StreamCut startStreamCut, final StreamCut endStreamCut) {
+           return stream(null, streamName, startStreamCut, endStreamCut);
```
Comment (fpj) [inline client/src/main/java/io/pravega/client/stream/ReaderGroupConfig.java]: You could pass the default scope name rather than null here.

File: client/src/test/java/io/pravega/client/stream/ReaderGroupConfigTest.java [hunk 8]
```diff
@@ -28,14 +28,21 @@
     public void testValidConfigWithScopedStreamName() {
         ReaderGroupConfig cfg = ReaderGroupConfig.builder()
                 .disableAutomaticCheckpoints()
-                .stream("scope/s1", getStreamCut("s1"))
-                .stream("scope/s2", getStreamCut("s2"))
+                .stream("scope", "s1", getStreamCut("s1"))
+                .stream("scope", "s2", getStreamCut("s2"))
+                .stream("s3", getStreamCut("s3"))
```
Comment (fpj) [inline client/src/test/java/io/pravega/client/stream/ReaderGroupConfigTest.java]: One quick clarification: aren't we supposed to pass a default scope somewhere in this configuration?

Comment (shrids) [reply client/src/test/java/io/pravega/client/stream/ReaderGroupConfigTest.java]: The default scope is passed while creating a `ReaderGroupManager` instance.
`io.pravega.client.admin.ReaderGroupManager#withScope(java.lang.String, java.net.URI)`

File: client/src/main/java/io/pravega/client/stream/ReaderGroupConfig.java (hunk 7 above)
Comment (shrids) [reply client/src/main/java/io/pravega/client/stream/ReaderGroupConfig.java]: Passing null because the ReaderGroupConfig does not know the default Scope for the ReaderGroup. It is passed while creating the ReaderGroupManager `io.pravega.client.admin.ReaderGroupManager#withScope(java.lang.String, java.net.URI)`

File: client/src/main/java/io/pravega/client/stream/ReaderGroupConfig.java (hunk 4 above)
Comment (shrids) [reply client/src/main/java/io/pravega/client/stream/ReaderGroupConfig.java]: Updated the code to use NameUtils

File: client/src/main/java/io/pravega/client/stream/ReaderGroupConfig.java (hunk 6 above)
Comment (shrids) [reply client/src/main/java/io/pravega/client/stream/ReaderGroupConfig.java]: removed it.

File: client/src/main/java/io/pravega/client/stream/ReaderGroupConfig.java (hunk 3 above)
Comment (shrids) [reply client/src/main/java/io/pravega/client/stream/ReaderGroupConfig.java]: updated the comment.

File: client/src/main/java/io/pravega/client/stream/ReaderGroupConfig.java (hunk 7 above)
Comment (fpj) [reply client/src/main/java/io/pravega/client/stream/ReaderGroupConfig.java]: The semantics here is confusing. If I understand it right, in the case the scope is null, we assume that the stream name is scoped as per line 76. But, if we have given a default scope to the reader group manager, then why are we not making use of it? In fact, it makes sense to have a default scope for the reader group, not the manager, unless the default scope of the manager is also propagated to the reader groups.

**Comment by shrids** [comment]
Will open a new PR with an alternative approach for this issue
